curvefs/src/client/rpcclient/metacache.cpp

```diff
@@ -0,0 +1,347 @@
+                     << "} do not find partition";
+        return false;
+    }
+
+    // get copysetInfo with (poolID, copysetID)
+    CopysetInfo<MetaserverID> targetInfo;
+    if (false ==
+        GetCopysetInfowithCopySetID(*poolID, *copysetID, &targetInfo)) {
+        LOG(WARNING) << "{fsid:" << fsID << "inodeid:" << inodeID
+                     << ", logicalpooid" << *poolID
+                     << ", copysetid:" << *copysetID
+                     << "} do not find copyset info";
+        return false;
+    }
+
+    if (false == refresh && false == targetInfo.LeaderMayChange()) {
+        if (0 == targetInfo.GetLeaderInfo(metaServerID, target)) {
+            return true;
+        }
+
+        LOG(WARNING) << "{fsid:" << fsID << "inodeid:" << inodeID
+                     << ", logicalpooid:" << *poolID
+                     << ", copysetid:" << *copysetID
+                     << "} get leader from cache fail."
+                     << " current leader:" << *metaServerID;
+    }
+
+    // when need refresh or leader change, we need to update copysetinfo
+    bool ret = true;
+    uint32_t retry = 0;
+    while (retry++ < metacacheopt_.metacacheGetLeaderRetry) {
+        // refresh from metaserver
+        ret = UpdateLeaderInternal(*poolID, *copysetID, &targetInfo);
+        if (ret) {
+            targetInfo.ResetSetLeaderUnstableFlag();
+            UpdateCopysetInfo(*poolID, *copysetID, targetInfo);
+            break;
+        } else {
+            LOG(INFO) << "refresh leader from metaserver failed, "
+                      << "get copyset peer list from mds, "
+                      << "logicpool id = " << *poolID
+                      << ", copyset id = " << *copysetID;
+        }
+
+        // refresh from mds
+        ret = UpdateCopysetInfoFromMDS(*poolID, *copysetID, &targetInfo);
+        if (ret) {
+            break;
+        }
+
+        bthread_usleep(metacacheopt_.metacacheRPCRetryIntervalUS);
+    }
+
+    if (!ret) {
+        LOG(WARNING) << "get leader failed after retry!"
+                     << ", copysetid = " << *copysetID
+                     << ", logicpoolid = " << *poolID;
+        return false;
+    }
+
+    *applyIndex = targetInfo.GetAppliedIndex();
+    return 0 == targetInfo.GetLeaderInfo(metaServerID, target);
+}
+
+
+bool MetaCache::SelectTarget(uint32_t fsID, MetaserverID *metaServerID,
+                             butil::EndPoint *target, LogicPoolID *poolID,
+                             CopysetID *copysetID, uint32_t *partitionID,
+                             uint64_t *applyIndex) {
+    // get partitionlist with fsid
+    PatitionInfoList pinfoList;
+    if (false == GetParitionListWihFsID(fsID, &pinfoList)) {
+        LOG(WARNING) << "select target for {fsid:" << fsID
+                     << "} fail, parition list not exist";
+        return false;
+    }
+
+    // random select a partition from list
+    int index = rand_r(&seed) % pinfoList.size();
```

Comment:
seed is a global variable, it makes `rand_r(&seed)` almost the same with `rand()` in multi-thread situtation, because `rand()` maintain a hidden seed for all threads.

you should define the seed as thread_local, and another problem is for all thread_local seeds, they must have different initial value, otherwise rand_r will produces random numbers in the same order

Reply:
![image](https://user-images.githubusercontent.com/15689619/132531557-993c2243-a716-4763-9c60-b090e95e32d2.png)

Threadsafe: No. rand() is not threadsafe, but rand_r() is.

Reply:
ok